### PR TITLE
ignore micro-version mismatch for databricks pyspark

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

ignore micro-version mismatch for databricks pyspark

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

ignore micro-version mismatch for databricks pyspark

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !

File: mlflow/utils/requirements_utils.py [hunk 1]
```diff
@@ -445,6 +445,12 @@ def _check_requirement_satisfied(requirement_str):
     req = pkg_resources.Requirement.parse(requirement_str)
     pkg_name = req.name
 
+    if pkg_name == 'pyspark' and is_in_databricks_runtime():
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:448]: Can we add a test?

File: mlflow/utils/requirements_utils.py [hunk 2]
```diff
@@ -445,6 +445,12 @@ def _check_requirement_satisfied(requirement_str):
...
+        # for databricks runtime pyspark, ignore micro-version mismatch.
+        # replace == specifier to be ~= specifier, but keep === specifier unchanged.
+        requirement_str = re.sub('([^=])==([^=])', r'\1~=\2', requirement_str)
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:451]: Can we run this operation after `_init_packages_to_modules_map`?

Comment (harupy) [reply mlflow/utils/requirements_utils.py:451]:
```suggestion
        requirement_str = re.sub(r'==([^=])', r'~=\1', requirement_str)
```

Can we only replace the version specifier?

Comment (WeichenXu123) [reply mlflow/utils/requirements_utils.py:451]: This seems does not work for `===`
`re.sub(r'==([^=])', r'~=\1', 'abc===def')` get `'abc=~=def'`

File: tests/pyfunc/test_warn_dependency_requirement_mismatches.py [hunk 3]
```diff
@@ -9,6 +9,18 @@
...
+def _gen_mock_get_installed_version_fn(mock_versions):
+    original_get_installed_version_fn = mlflow.utils.requirements_utils._get_installed_version
+
+    def mock_get_installed_version_fn(package, module=None):
+        if package in mock_versions:
+            return mock_versions[package]
+        else:
+            return original_get_installed_version_fn(package, module)
+
+    return mock_get_installed_version_fn
```
Comment (harupy) [inline tests/pyfunc/test_warn_dependency_requirement_mismatches.py:21]:
```python
@contextmanager
def mock_get_installed_version(mock_versions):
    original_get_installed_version = mlflow.utils.requirements_utils._get_installed_version

    def new_get_installed_version(package, module=None):
        if package in mock_versions:
            return mock_versions[package]
        else:
            return original_get_installed_version(package, module)

    with mock.patch(
         "mlflow.utils.requirements_utils._get_installed_version",
         mocked_get_installed_version) as mock_get_installed_version:
         yield mock_get_installed_version
```

Can we define a context manager like this?

File: mlflow/utils/requirements_utils.py [hunk 4]
```diff
@@ -442,6 +442,12 @@ def _check_requirement_satisfied(requirement_str):
     Otherwise, returns an instance of `_MismatchedPackageInfo`.
     """
     _init_packages_to_modules_map()
+
+    if re.match(r"pyspark[^\w-]", requirement_str) and is_in_databricks_runtime():
+        # for databricks runtime pyspark, ignore micro-version mismatch.
+        # replace == specifier to be ~= specifier, but keep === specifier unchanged.
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:448]:
```suggestion
        # replace == operator with ~= operator, but keep === operator unchanged.
```

File: tests/pyfunc/test_warn_dependency_requirement_mismatches.py [hunk 5]
```diff
@@ -110,3 +111,30 @@ def mock_get_installed_version_fn(package, module=None):
...
             )
+
+
+@pytest.mark.large
+def test_warn_dependency_requirement_mismatches_ignore_databricks_runtime_micro_version(tmpdir):
+    req_file = tmpdir.join("requirements.txt")
+    req_file.write("pyspark==3.2.1")
+
+    with mock.patch("mlflow.pyfunc._logger.warning") as mock_warning:
+        with mock.patch("mlflow.utils.requirements_utils.is_in_databricks_runtime", lambda: True):
```
Comment (harupy) [inline tests/pyfunc/test_warn_dependency_requirement_mismatches.py:122]:
```suggestion
        with mock.patch("mlflow.utils.requirements_utils.is_in_databricks_runtime", return_value=True):
```

you could also use `return_value`

File: tests/pyfunc/test_warn_dependency_requirement_mismatches.py [hunk 6]
```diff
@@ -110,3 +111,30 @@ def mock_get_installed_version_fn(package, module=None):
...
+
+        with mock.patch(
+            "mlflow.utils.requirements_utils._get_installed_version",
+            _gen_mock_get_installed_version_fn({"pyspark": "3.2.2"}),
+        ):
+            _warn_dependency_requirement_mismatches(model_path=tmpdir)
+            mock_warning.assert_called_once_with(
+                AnyStringWith(" - pyspark (current: 3.2.2, required: pyspark==3.2.1)")
+            )
+            mock_warning.reset_mock()
```
Comment (harupy) [inline tests/pyfunc/test_warn_dependency_requirement_mismatches.py:140]: `3.2.2` and `3.2.1` should be considered as matched, right?

File: mlflow/utils/requirements_utils.py (hunk 2 above)
Comment (harupy) [reply mlflow/utils/requirements_utils.py]: Got it, I misunderstood `===` was a typo, but it's a valid operator.

File: tests/pyfunc/test_warn_dependency_requirement_mismatches.py (hunk 6 above)
Comment (WeichenXu123) [reply tests/pyfunc/test_warn_dependency_requirement_mismatches.py:140]: this is test for non-databricks pyspark.

File: mlflow/utils/requirements_utils.py [hunk 7]
```diff
@@ -442,6 +442,12 @@ def _check_requirement_satisfied(requirement_str):
     Otherwise, returns an instance of `_MismatchedPackageInfo`.
     """
     _init_packages_to_modules_map()
+
+    if re.match(r"pyspark[^\w-]", requirement_str) and is_in_databricks_runtime():
+        # for databricks runtime pyspark, ignore micro-version mismatch.
+        # replace == operator with ~= operator, but keep === operator unchanged.
+        requirement_str = re.sub("([^=])==([^=])", r"\1~=\2", requirement_str)
```
Comment (harupy) [inline mlflow/utils/requirements_utils.py:449]: Can this handle a requirement string that looks like `pyspark>=1.2.3,==1.2.*`?

Comment (WeichenXu123) [reply mlflow/utils/requirements_utils.py:449]: emm. this is an issue.

Comment (WeichenXu123) [reply mlflow/utils/requirements_utils.py:449]: i will update regex to cover the case ==3.2 or ==3.2.*

Comment (dbczumar) [reply mlflow/utils/requirements_utils.py:449]: I don't think it needs to, right? The purpose of this change is to make sure that exact requirement specifiers generated automatically by MLflow, e.g. `pyspark==3.2.1`, are not interpreted too strictly on Databricks. If the user wants to supply a custom / stricter specifier, then they should be able to without us interfering / interpreting the specifier differently.

Comment (harupy) [reply mlflow/utils/requirements_utils.py:449]: I'm trying to come up with a more robust approach:

```python
import pkg_resources
from packaging.requirements import Requirement
from packaging.specifiers import SpecifierSet
import re


def is_satisfied(req, version):
    req = pkg_resources.Requirement.parse(req)
    if req.name == "pyspark" and len(req.specifier) == 1:
        m = re.match(r"==([^=]+)", str(req.specifier))
        if m:
            return SpecifierSet(f"~={m.group(1)}").contains(version)
    return req.specifier.contains(version)


print(is_satisfied("pyspark == 1.2.2", "1.2.3"))  # True
print(is_satisfied("pyspark>=1.2.3,==1.2.*", "1.2.3"))  # True
print(is_satisfied("pyspark === abc", "1.2.3"))  # False
print(is_satisfied("scikit-learn == 1.2.2", "1.2.3"))  # False

```

Comment (dbczumar) [reply mlflow/utils/requirements_utils.py:449]: Actually, @WeichenXu123 , do we need this PR? The micro version isn't going to change within the same MLR / DBR version, right?

**Review by dbczumar — CHANGES_REQUESTED** [review]
On second thought, not sure we need to do this. See https://github.com/mlflow/mlflow/pull/5386/files#r808837619

File: mlflow/utils/requirements_utils.py (hunk 7 above)
Comment (WeichenXu123) [reply mlflow/utils/requirements_utils.py:449]: Confirmed with @hyukjin  the pyspark version is not changed within the same MLR/DBR version

**Pull request closed without merging**